### ISSUE TITLE
Fix crash with latest python3-tk (v3.8.2) in debian bullseye

### ISSUE THREAD

**Issue by pav-rez** [frougon/FFGo#2]
Dear maintainer,

  a recent update of the python3-tk (in Debian bullseye) to version 3.8.2 is causing ffgo to crash. Attached is a patch solving the problem.

Best regards,
Pavel

[ffgo_python3tk.patch.txt](https://github.com/frougon/FFGo/files/4529039/ffgo_python3tk.patch.txt)

**Comment by frougon** [comment]
Hi Pavel,

Many thanks for your patch! It's been a long time since I last touched this code. I see that FFGo indeed doesn't work anymore on Debian unstable. This looks like a Tkinter regression to me, or a packaging bug. Do you have more info? In particular, do you know if this regression (from the looks of it) has been publically discussed?

More importantly, can you say why your patch fixes the problem and if it will work fine with previous versions of Tkinter? Is my code doing something incorrect that would explain the bad behavior?

Sorry for all these questions, but since your patch looks a bit mysterious to me, I need to understand it better. :-)

Thanks again, and take care!

**Comment by pav-rez** [comment]
Dear Florent,

  firstly, I have to apologize - I am very far from being a python expert. It is thus possible (likely) that while the patch works, it may not be a proper one... The patch was found doing a bit of "experimental" programming while staring at the latest python3-tk code. I have now revised it a bit (see attachment), switching from internal Treeview._selection method to the proper Treeview.selection_set method.

The origin of the problem is simple: python3-tk in python3.7 has a method (in ttk.py):

Treeview.selection(self, selop=_sentinel, items=None)

where selop is an option-string defining what kind of selection is being done (in [FFGo/gui/widgets.py](https://github.com/frougon/FFGo/blob/master/ffgo/gui/widgets.py#L112) code the selop="set"). However, already in python3.7 there is a warning that the selection method with the selop option is being [deprecated](https://github.com/python/cpython/commit/97f1ca167385bafd9be5d226f61c33412fc081bf#diff-9aa8b20d5d2878ccf22224ead3f3af5a).

The warning turned into reality in python3-tk in python3.8, replacing the single selection method by a set of methods [selection_set/add/remove/toogle](https://github.com/python/cpython/blob/master/Lib/tkinter/ttk.py#L1425).

The patch I made should be backward compatible - trying firstly the deprecated method and if it fails it turns to the new selection_set method.

Best regards,
Pavel

[ffgo_python3.8tk.patch.v2.txt](https://github.com/frougon/FFGo/files/4532675/ffgo_python3.8tk.patch.v2.txt)

**Comment by frougon** [comment]
Many thanks for the links and new patch, this is really helpful! This patch looks better to me. I tested it but still have a problem: if I choose a non-existent aircraft in the aircraft list or a non-existent airport in the airport list or in the GPS Tool, the assertion from line 672 (after your second patch has been applied) is triggered. So, there is still a behavioral change that needs investigation. This is the assertion in question:

```
    def onTreeviewSelect(self, event=None):
        tree = self.treeWidget
        currentSel = tree.selection()
        assert currentSel, "Unexpected empty selection in TreeviewSelect event"
```

If you can explain the change in behavior, that's great; otherwise, I'll need to find a bit more time to dive into this code again. :-)

I'm rather disappointed to see how Ttk broke compatibility here...

Regards

**Comment by pav-rez** [comment]
Well, I can't explain the change in the behavior. I have to admit I do not remember what was actually the original correct behavior for these cases of selecting a non-existent aircraft or airport. However, the only solution to go around the problem I have found was to completely avoid these cases: by offering a full un-filtered list of aircrafts (airports) in case no match is found - see attached new patch.

Obviously, this is just a workaround, not a fix - but I haven't found a way for a proper fix...

Best regards.
[ffgo_python3.8tk.patch.v3.txt](https://github.com/frougon/FFGo/files/4536458/ffgo_python3.8tk.patch.v3.txt)

**Comment by frougon** [comment]
Thanks for your input. IIRC, the behavior was to display an empty list (no match => no entry to show; then, as you erase chars and matches start to exist, the list grows). I'll try to find some time in the next days (hopefully) to better investigate this issue.

Regards and take care. :-)

**Comment by pav-rez** [comment]
I see. There is actually no problem to achieve displaying empty list (without a crash), just by replacing the assertion from line 672 with a condition:

```
if len(currentSel) == 0:
    self.setNullOutputVar()
    return
```

However, while FFGo correctly shows that it will simply not use any aircraft argument when running fgfs, there is an error message (no python crash) from FFGo like "No such aircraft '' in ''" or "No such airport". Both these error messages looked like a piece of debugging information to me, rather than something intended for an ordinary user. This is why I was hesitating about the intended behavior...

**Comment by frougon** [comment]
Hi,

Sorry for the delay. I reread widgets.py and tested on a Debian buster system. As expected, the bug doesn't exist on buster. No `<<TreeviewSelect>>` even was generated with buster's Ttk when the selection became empty. I think the most faithful way to keep the behavior we used to have is to replace

```python
self.setOutputVarForItem(currentSel[0])
```

with

```python
if currentSel:
    self.setOutputVarForItem(currentSel[0])
```

at the end of `IncrementalChooser.onTreeviewSelect()` (plus the `try... except TypeError: ...` you proposed earlier to cater for the incompatible changes in `Treeview.selection()`). The advantage as compared to doing `self.setNullOutputVar()` is that if one types a search string and at some point one does a typing mistake that renders the list of matches empty, then after using backspace enough times to correct the mistake, the selection gets back to the item that was selected just before the list of matches became empty.

With this, there *are* the error messages you mentioned if one clicks “Run FG” while the list of airports or aircraft is empty. This is indeed not perfect, but a rather harmless corner case that was already like this before the incompatible changes in Ttk were introduced (I verified this on buster; I probably simply hadn't tried this situation when writing the code).

I've pushed a fix for the bug reported here as proposed above (commit 72e4554b8c3dc180ddb911ebe51f8d49a709a01a) and will look later at the harmless corner cases, as time permits. Thanks for your report!

Regards
